app/models/batch_update_form/contractor.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+class BatchUpdateForm::Contractor < BatchUpdateForm::Base
+  model_class 'Contractor'
+  attribute :enabled, type: :boolean
+  attribute :vendor, type: :boolean
+  attribute :customer, type: :boolean
+  attribute :description
+  attribute :address
+  attribute :phones
+  attribute :smtp_connection_id, type: :foreign_key, class_name: 'System::SmtpConnection'
+
+  validates :vendor, required_with: :customer
+
+  validate if: -> { customer_changed? || vendor_changed? } do
+    if customer == 'f' && vendor == 'f'
+      errors.add :base, I18n.t('activerecord.errors.models.contractor.vendor_and_customer')
+    elsif customer == 't'
+      errors.add :base, I18n.t('activerecord.errors.models.contractor.vendor_and_customer') if vendor == 't'
+    elsif vendor == 't'
+      errors.add :base, I18n.t('activerecord.errors.models.contractor.vendor_and_customer') if customer == 't'
+    end
+  end
+
+  # recommended to put this heavy validation latest
+  validate :without_customers_auths
+
+  def without_customers_auths
+    if errors.messages.blank? && (customer == 'f' && vendor == 't')
+      errors.add(:customer, I18n.t('activerecord.errors.models.contractor.attributes.customer')) if customers_auth_any?
+    end
+  end
+
+  def customers_auth_any?
+    if selected_record.any?
+      selected_record.each do |id|
+        return true if Contractor.find_by(id: id)&.customers_auths&.any?
+      end
+      false
+    else
+      # find_each retrieves Contractors in batches of 1000 and yields them to the block one by one
+      Contractor.find_each do |contractor|
+        return true if contractor.customers_auths.any?
+      end
+      false
+    end
```

Comment:
this code have to much problems (complex, n+1)
def customers_auth_any?
  CustomersAuth.where(customer_id: selected_record || Contractor.select(:id)).any?
  end